src/coord/rmr/redis_cluster.c

```diff
@@ -13,110 +13,174 @@
   RS_AutoMemory(ctx);
-  const char *myId = NULL;
-  RedisModuleCallReply *r = RedisModule_Call(ctx, "CLUSTER", "c", "MYID");
-  if (r == NULL || RedisModule_CallReplyType(r) != REDISMODULE_REPLY_STRING) {
+  RedisModuleCallReply *myID_reply = RedisModule_Call(ctx, "CLUSTER", "c", "MYID");
+  if (myID_reply == NULL || RedisModule_CallReplyType(myID_reply) != REDISMODULE_REPLY_STRING) {
     RedisModule_Log(ctx, "warning", "Error calling CLUSTER MYID");
     return NULL;
   }
+
   size_t idlen;
-  myId = RedisModule_CallReplyStringPtr(r, &idlen);
+  const char *myID = RedisModule_CallReplyStringPtr(myID_reply, &idlen);
 
-  r = RedisModule_Call(ctx, "CLUSTER", "c", "SLOTS");
-  if (r == NULL || RedisModule_CallReplyType(r) != REDISMODULE_REPLY_ARRAY) {
-    RedisModule_Log(ctx, "warning", "Error calling CLUSTER SLOTS");
+  RedisModuleCallReply *cluster_shards = RedisModule_Call(ctx, "CLUSTER", "c", "SHARDS");
+  if (cluster_shards == NULL || RedisModule_CallReplyType(cluster_shards) != REDISMODULE_REPLY_ARRAY) {
+    RedisModule_Log(ctx, "warning", "Error calling CLUSTER SHARDS");
     return NULL;
   }
-
-  /*1) 1) (integer) 0
-   2) (integer) 5460
-   3) 1) "127.0.0.1"
-      2) (integer) 30001
-      3) "09dbe9720cda62f7865eabc5fd8857c5d2678366"
-   4) 1) "127.0.0.1"
-      2) (integer) 30004
-      3) "821d8ca00d7ccf931ed3ffc7e3db0599d2271abf"*/
-
-  size_t len = RedisModule_CallReplyLength(r);
-  if (len < 1) {
-    RedisModule_Log(ctx, "warning", "Got no slots in CLUSTER SLOTS");
+  /*
+1) 1# "slots" =>
+      1) (integer) 0
+      2) (integer) 4095
+      3) (integer) 8192
+      4) (integer) 12287
+   2# "nodes" =>
+      1)  1# "id" => "e10b7051d6bf2d5febd39a2be297bbaea6084111"
+          2# "port" => (integer) 30001
+          3# "tls-port" => (integer) 40001
+          4# "ip" => "127.0.0.1"
+          5# "endpoint" => "localhost"
+          6# "role" => "master"
+      2)  1# "id" => "821d8ca00d7ccf931ed3ffc7e3db0599d2271abf"
+          2# "port" => (integer) 30004
+          3# "tls-port" => (integer) 40004
+          4# "ip" => "127.0.0.1"
+          5# "endpoint" => "localhost"
+          6# "role" => "replica"
+2) 1# "slots" =>
+      1) (integer) 4096
+      2) (integer) 8191
+      3) (integer) 12288
+      4) (integer) 16383
+   2# "nodes" =>
+      1)  1# "id" => "fd20502fe1b32fc32c15b69b0a9537551f162f1f"
+          2# "port" => (integer) 30003
+          3# "tls-port" => (integer) 40003
+          4# "ip" => "127.0.0.1"
+          5# "endpoint" => "localhost"
+          6# "role" => "master"
+      2)  1# "id" => "6daa25c08025a0c7e4cc0d1ab255949ce6cee902"
+          2# "port" => (integer) 30005
+          3# "tls-port" => (integer) 40005
+          4# "ip" => "127.0.0.1"
+          5# "endpoint" => "localhost"
+          6# "role" => "replica"
+  */
+
+  size_t numShards = RedisModule_CallReplyLength(cluster_shards);
+  if (numShards == 1 && hasNoSlots(RedisModule_CallReplyArrayElement(cluster_shards, 0))) {
```

Comment:
is this even possible?

Reply:
Is zero length response still possible?

Reply:
1. YES - if the cluster is not set correctly. While `CLUSTER SLOTS` returns no slots in this case, `CLUSTER SHARDS` always returns some info about the self.
2. I don't think it's possible, but I'll make this condition fail in this case as well